controlpanel/api/cluster.py

```diff
@@ -346,3 +356,120 @@ def get_repository(user, repo_name):
         return github.get_repo(repo_name)
     except GithubException.UnknownObjectException:
         return None
+
+
+class ToolDeploymentError(Exception):
+    pass
+
+
+def deploy_tool(tool, user, **kwargs):
+    values = {
+        'username': user.username.lower(),
+        'Username': user.username.lower(),  # XXX backwards compatibility
+        'aws.iamRole': user.iam_role_name,
+        'toolsDomain': settings.TOOLS_DOMAIN,
+    }
+
+    # generate per-deployment secrets
+    for key, value in tool.values.items():
+        if value == '<SECRET_TOKEN>':
+            tool.values[key] = secrets.token_hex(32)
+
+    values.update(tool.values)
+    values.update(kwargs)
+    values = ','.join(f'{key}={val}' for key, val in values.items())
+
+    try:
+        return helm.upgrade_release(
+            f'{tool.chart_name}-{user.slug}',
+            f'mojanalytics/{tool.chart_name}',  # XXX assumes repo name
+            f'--version', tool.version,
+            f'--namespace', user.k8s_namespace,
+            f'--set', values,
+        )
+
+    except HelmError as error:
+        raise ToolDeploymentError(error)
+
+
+def list_tool_deployments(user, search_name=None, search_version=None):
+    deployments = []
+    k8s = KubernetesClient()
+    results = k8s.AppsV1Api.list_namespaced_deployment(user.k8s_namespace)
+    for deployment in results.items:
+        app_name = deployment.metadata.labels["app"]
```

Comment:
Shame we don't have some label with the helm chart name + version in our helm charts 😞  - Helm documentation mentions this: [The Chart Best Practices Guide - Labels and Annotations](https://helm.sh/docs/chart_best_practices/#labels-and-annotations).

Actually, it seems that at least the latest versions of the tools have the `chart` label which is something like `chart: "{{ .Chart.Name }}-{{ .Chart.Version }}"`. I just did a bit of git digging and obvs Jupyter and Airflow are copied from RStudio chart and RStudio had that label for 2 years so maybe it's safe to use the `chart` label after all - it may be less brittle than use the `app` label.

Reply:
Oh hang on, you do use the `chart` label, sorry 🤦‍♂